sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java

```diff
@@ -0,0 +1,446 @@
+
+      /**
+       * The underlying {@link FileIO.Write} that writes converted input to CSV formatted output.
+       */
+      abstract Builder<T> setTextIOWrite(TextIO.Write value);
+
+      /** The {@link CSVFormat} to convert input. Defaults to {@link CSVFormat#DEFAULT}. */
+      abstract Builder<T> setCSVFormat(CSVFormat value);
+
+      abstract CSVFormat getCSVFormat();
+
+      abstract Write<T> autoBuild();
+
+      final Write<T> build() {
+        checkArgument(
+            !getCSVFormat().getSkipHeaderRecord(),
+            "withSkipHeaderRecord is an illegal CSVFormat setting");
+
+        if (getCSVFormat().getHeaderComments() != null) {
+          checkArgument(
+              getCSVFormat().isCommentMarkerSet(),
+              "CSVFormat withCommentMarker required when withHeaderComments");
+        }
+        return autoBuild();
+      }
+    }
+
+    @Override
+    public WriteFilesResult<String> expand(PCollection<T> input) {
+      if (!input.hasSchema()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "%s requires an input Schema. Note that only Row or user classes are supported. Consider using TextIO or FileIO directly when writing primitive types",
+                Write.class.getName()));
+      }
+
+      Schema schema = input.getSchema();
+
+      RowCoder rowCoder = RowCoder.of(schema);
+
+      PCollection<Row> rows =
+          input
+              .apply("To Rows", MapElements.into(rows()).via(input.getToRowFunction()))
+              .setCoder(rowCoder);
+
+      CSVFormat csvFormat = applyRequiredCSVFormatSettings(schema);
+
+      String header = formatHeader(csvFormat);
+
+      SerializableFunction<Row, String> toCsvFn =
+          CsvRowConversions.RowToCsv.builder()
+              .setCSVFormat(csvFormat)
+              .setSchema(input.getSchema())
+              .build();
+
+      PCollection<String> csv = rows.apply("To CSV", MapElements.into(strings()).via(toCsvFn));
+
+      return csv.apply("Write CSV", getTextIOWrite().withHeader(header).withOutputFilenames());
+    }
+
+    CSVFormat applyRequiredCSVFormatSettings(Schema schema) {
+      CSVFormat csvFormat = getCSVFormat().withSkipHeaderRecord();
+      if (csvFormat.getHeader() == null) {
+        csvFormat = csvFormat.withHeader(schema.sorted().getFieldNames().toArray(new String[0]));
+      }
+      return csvFormat;
+    }
+
+    private static String formatHeader(CSVFormat csvFormat) {
+      String[] header = requireNonNull(csvFormat.getHeader());
+      List<String> result = new ArrayList<>();
+      if (csvFormat.getHeaderComments() != null) {
+        for (String comment : csvFormat.getHeaderComments()) {
+          result.add(csvFormat.getCommentMarker() + " " + comment);
+        }
+      }
+      CSVFormat withoutHeaderComments = csvFormat.withHeaderComments();
+      result.add(withoutHeaderComments.format((Object[]) header));
+      return String.join("\n", result);
+    }
```

Comment:
We are ignoring the case where the user wants to skip the header record and doesn't specify the header. We should only specify the `csvFormat` header if `withSkipHeaderRecord` is `false`.

Note that we must also return `null` from formatHeader in this case since `TextIO#withHeader` specifically says that a newline is automatically added after.

Add a test case for this scenario.

```suggestion
    CSVFormat applyRequiredCSVFormatSettings(Schema schema) {
      CSVFormat csvFormat = getCSVFormat();
      if (!csvFormat.getSkipHeaderRecord() && csvFormat.getHeader() == null) {
        return csvFormat.withSkipHeaderRecord().withHeader(schema.sorted().getFieldNames().toArray(new String[0]));
      }
      return csvFormat;
    }

    private static @Nullable String formatHeader(CSVFormat csvFormat) {
      String[] header = requireNonNull(csvFormat.getHeader());
      if (header == null) {
        return null;
      }
      List<String> result = new ArrayList<>();
      if (csvFormat.getHeaderComments() != null) {
        for (String comment : csvFormat.getHeaderComments()) {
          result.add(csvFormat.getCommentMarker() + " " + comment);
        }
      }
      CSVFormat withoutHeaderComments = csvFormat.withHeaderComments();
      result.add(withoutHeaderComments.format((Object[]) header));
      return String.join("\n", result);
    }
```

Reply:
Overall, I changed this method to incorporate applying to TextIO.Write.withHeader to provide more clarity.

> We are ignoring the case where the user wants to skip the header record and doesn't specify the header. We should only specify the csvFormat header if withSkipHeaderRecord is false.
CSVFormat communicates that withSkipHeaderRecord functions to not output the header.  Do you agree then we should keep withSkipHeaderRecord separate from where we are deriving the header?

> Note that we must also return null from formatHeader in this case since TextIO#withHeader specifically says that a newline is automatically added after.
Are you saying that the consequence of TextIO#withHeader is:

```
col1, col2, col3

1,2,3
2,3,4
```

vs

```
col1, col2, col3
1,2,3
2,3,4
```

Reply:
I'm saying that when header is `null` the output is:
```
1,2,3
2,3,4
```
while when header is `""` the output is:
```

1,2,3
2,3,4
```

Reply:
I would like to anchor our discussion on CSV header-related software behavior using the following:

```gherkin
Feature: write a CSV header
    Scenario: User provides CSVFormat.withHeader
        Given CSVFormat is not null
        When CsvIO writes PCollection<Row> to CSV formatted file shards
        Then for each shard the CSVFormat withHeader appears as the first line

    Scenario: User does not provide CSVFormat.withHeader
        Given CSVFormat is null
        When CsvIO writes PCollection<Row> to CSV formatted file shards
        Then for each shard the Schema field names in sorted order appears as the first line

    Scenario: User provides CSVFormat.withSkipHeaderRecord == true
        Given CSVFormat.withSkipHeaderRecord is true
        When the user applies the CsvIO.Write PTransform
        Then they see an error that this is an illegal setting
```